sycl/test-e2e/DeviceDependencies/free_function_kernels.cpp

```diff
@@ -0,0 +1,270 @@
+// Ensure -fsycl-allow-device-dependencies can work with free function kernels.
+
+// REQUIRES: aspect-usm_shared_allocations
+// RUN: %{build} -o %t.out -fsycl-allow-device-dependencies
+// RUN: %{run} %t.out
+
+// The name mangling for free function kernels currently does not work with PTX.
+// UNSUPPORTED: cuda
+
+#include <iostream>
+#include <sycl/detail/core.hpp>
+#include <sycl/ext/oneapi/free_function_queries.hpp>
+#include <sycl/usm.hpp>
+
+using namespace sycl;
+
+void printUSM(int *usmPtr, int size) {
+  std::cout << "usmPtr[] = {";
+  for (int i = 0; i < size; i++) {
+    std::cout << usmPtr[i] << ", ";
+  }
+  std::cout << "}\n";
+}
+
+bool checkUSM(int *usmPtr, int size, int *Result) {
+  bool Pass = true;
+  for (int i = 0; i < size; i++) {
+    if (usmPtr[i] != Result[i]) {
+      Pass = false;
+      break;
+    }
+  }
+  if (Pass)
+    return true;
+
+  std::cout << "Expected = {";
+  for (int i = 0; i < size; i++) {
+    std::cout << Result[i] << ", ";
+  }
+  std::cout << "}\n";
+  std::cout << "Result   = {";
+  for (int i = 0; i < size; i++) {
+    std::cout << usmPtr[i] << ", ";
+  }
+  std::cout << "}\n";
+  return false;
+}
+
+extern "C" SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
+    (ext::oneapi::experimental::single_task_kernel)) void ff_0(int *ptr,
+                                                               int start,
+                                                               int end) {
+  for (int i = start; i <= end; i++)
+    ptr[i] = start + end;
+}
+
+bool test_0(queue Queue) {
+  constexpr int Range = 10;
+  int *usmPtr = malloc_shared<int>(Range, Queue);
+  int start = 3;
+  int end = 5;
+  int Result[Range] = {0, 0, 0, 8, 8, 8, 0, 0, 0, 0};
+  range<1> R1{Range};
+
+  memset(usmPtr, 0, Range * sizeof(int));
+  Queue.submit([&](handler &Handler) {
+    Handler.single_task([=]() {
+      for (int i = start; i <= end; i++)
+        usmPtr[i] = start + end;
+    });
+  });
+  Queue.wait();
+  bool PassA = checkUSM(usmPtr, Range, Result);
+  std::cout << "Test 0a: " << (PassA ? "PASS" : "FAIL") << std::endl;
+
+  bool PassB = false;
+#ifndef __SYCL_DEVICE_ONLY__
```

Comment:
Unless that's a known issue with free function kernels that requires use to use `__SYCL_DEVICE_ONLY__` you should not be using such `#ifdef`s.

Reply:
Fine by me to defer a resolution to a separate PR, but it would be good to leave some kind of comment/`TODO` maybe to note that

Reply:
Add TODO comment